### PR TITLE
Revised Notifers with added Subscriptions

### PR DESCRIPTION
**Pull request by tyg**

### CODE REVIEW

**Comment by katelynsills** [comment]
@tyg, can we get @erights or someone else to review this? @Chris-Hibbert and I aren't the experts on this one, and we have a real crunch this week.

**Comment by tyg** [comment]
I can add him. I was actually thinking more Chris as this was his thing,
and knowing of your crunch only included you since you'd indicated you
wanted to at least be on the review list for any major doc changes not
previously OKed by you not to do so.

Tom

**Comment by katelynsills** [comment]
> I can add him. I was actually thinking more Chris as this was his thing, and knowing of your crunch only included you since you'd indicated you wanted to at least be on the review list for any major doc changes not previously OKed by you not to do so. Tom

Thanks, I appreciate it! I think it was not @Chris-Hibbert  and was actually @kriskowal and @erights who build the notifier and subscription, but I could be wrong.

**Comment by Chris-Hibbert** [comment]
I did write the original version of the notifier, but that was with extensive direction from @erights. From that point, I was in the forefront of ensuring it got documented thoroughly. Once asyncIterators and subscription were introduced, @kriskowal and @erights were very much responsible.  I'm no longer an expert, though I may be the one who has made the most use of it.

**Comment by tyg** [comment]
Ah. Added Kris then. Apologies to Chris; you'd initiated revising the docs on this so I'd thought you'd been responsible for the code.

File: main/guides/js-programming/notifiers.md [hunk 1]
```diff
@@ -1,94 +1,330 @@
...
+
+    If you call `getUpdateSince(oldUpdateCount)` with no count, or any 
+    `updateCount` other than the most recent one, the notifier immediately 
+    returns a promise for a record with the current state. If you call with 
+    the most-recently generated `updateCount`, the notifier returns a promise 
+    for the next record, which is resolved on the next state change. If you 
+    haven't called `getUpdateSince()` before, you won't have a 
+    previous `updateCount` to use.
+
+## Notifiers and Subscriptions in Zoe
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:149]: My review ends at Notifiers and Subscriptions in Zoe. Everything above looks _fantastic_.

File: main/guides/js-programming/notifiers.md [hunk 2]
```diff
@@ -1,94 +1,330 @@
...
+seats.
+
+Zoe's updates for an offer show the current allocation that will be paid if the contract completes without further changes.
+
+Individual contracts can use notifiers and subscriptions to provide updates giving current prices or other contract-specific details.
+
+The following methods use or return notifiers. Click on the name to go to their
+full documentation:
+
+- [`ZCFSeat.getNotifier()](/zoe/api/zoe-contract-facet.md#zcfseat-object)
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:167]:
```suggestion
- [`ZCFSeat.getNotifier()`](/zoe/api/zoe-contract-facet.md#zcfseat-object)
```

File: main/guides/js-programming/notifiers.md [hunk 3]
```diff
@@ -1,94 +1,330 @@
...
+  - Part of the Zoe API, returns a notifier associated with the seat. Its updates can be anything the contract wants to publish, such as
+     price changes, new currency pools, etc.
+- [`purse.getCurrentAmountNotifier()`](/ertp/api/purse.md#purse-getcurrentamountnotifier)
+   - Part of the ERTP API, returns a lossy notifier for changes to this purse's balance.
+- [`getPursesNotifier`](/guides/wallet/api.md#getpursesnotifier)
+   - Part of the Wallet API, it returns a notifier that follows changes in the purses in the Wallet.
+- [`getOffersNotifier`](/guides/wallet/api.md#getoffersnotifier)
+   - Part of the Wallet API, it returns a notifier that follows changes to the offers received by the Wallet.
+- [`makeQuoteNotifier(amountIn,brandOut)`](/repl/priceAuthority.md#makequotenotifier-amountin-brandout)
+   - Part of the PriceAuthority API, notifies the latest PriceQuotes for the given amountIn.
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:180]:
```suggestion
   - Part of the PriceAuthority API, notifies the latest `PriceQuotes` for the given `amountIn`.
```

File: main/guides/js-programming/notifiers.md [hunk 4]
```diff
@@ -1,94 +1,330 @@
...
+or publication/subscription pair on the producer's site. As a result, both the
+`iterationObserver` and the initial `asyncIterable` are on the producer's site. If
+Producer Paula sends Consumer Bob the `asyncIterable`, Bob receives a possibly
+remote reference to it. Producers and their Consumers can be remote from each
+other. 
+
+Bob's example code above is still correct if he uses this reference directly, since
+`observeIteration` only needs its first argument to be a reference of some sort to
+an `AsyncIterable` conveying `Passable` values. This reference may be a local
+`AsyncIterable`, a remote presence of an `AsyncIterabl`e, or a local or remote
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:280]:
```suggestion
`AsyncIterable`, a remote presence of an `AsyncIterable`, or a local or remote
```

File: main/guides/js-programming/notifiers.md [hunk 5]
```diff
@@ -1,94 +1,330 @@
...
+promise for an `AsyncIterable`. `observeIteration` only sends it eventual messages
+using `E()`  and so doesn't care about these differences.  
+
+However, Bob’s code is sub-optimal. Its distributed systems properties are
+not terrible, but Bob does better using `getSharableSubscriptionInternals()`
+(provided by `SubscriptionKit`). This lets Bob make a local `AsyncIterabl`e that
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:286]:
```suggestion
(provided by `SubscriptionKit`). This lets Bob make a local `AsyncIterable` that
```

File: main/guides/js-programming/notifiers.md [hunk 6]
```diff
@@ -1,94 +1,330 @@
...
+coordinates better with producer Paula's `IterationObserver`.  
+
+Subscriber Alice's above code is less forgiving. She's using JavaScript's
+for-await-of loop which requires a local `AsyncIterable`. It cannot handle a
+remote reference to an `AsyncIterable` at Paula's site. Alice has to make an
+`AsyncIterabl`e at her site by using `getSharableSubsciptionInternals()`. She can
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:292]:
```suggestion
`AsyncIterable` at her site by using `getSharableSubsciptionInternals()`. She can
```

File: main/guides/js-programming/notifiers.md [hunk 7]
```diff
@@ -1,94 +1,330 @@
...
+const localSubscription =
+  makeSubscription(E(subscription).getSharableSubsciptionInternals());
+  consume(localSubscription);
+```
+The above used a SubscriptionKit. NotifierKits have a similar pair of a
+`getSharableNotifierInternals()` method and a `makeNotifier`. However, this
+technique requires Alice know what kind of possibly-remote `AsyncIterable`
+she has, and to have the required making function code locally available. 
+
+Alternatively, Alice can generically mirror any possibly remote `AsyncIterabl`e by
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:306]:
```suggestion
Alternatively, Alice can generically mirror any possibly remote `AsyncIterable` by
```

File: main/guides/js-programming/notifiers.md [hunk 8]
```diff
@@ -1,94 +1,330 @@
...
+making a new local pair and plugging them together with `observeIteration`. 
+```js
+const {
+  publication: adapterPublication,
+  subscription: adapterSubscription
+} = makeSubscriptionKit();
+observeIteration(subscription, adapterPublication);
+consume(adapterSubscription);
+```
+This works when subscription is a reference to any `AsyncIterabl`e. If Alice only
```
Comment (kriskowal) [inline main/guides/js-programming/notifiers.md:316]:
```suggestion
This works when subscription is a reference to any `AsyncIterable`. If Alice only
```

**Review by kriskowal — APPROVED** [review]
Reviewed through Zoe to the end now.  👍

**Pull request merged**